pulsar/consumer_partition.go

```diff
@@ -2195,6 +2226,24 @@ func (pc *partitionConsumer) discardCorruptedMessage(msgID *pb.MessageIdData,
 }
 
 func (pc *partitionConsumer) hasNext() bool {
+
+	// If a seek by time has been performed, then the `startMessageId` becomes irrelevant.
+	// We need to compare `markDeletePosition` and `lastMessageId`,
+	// and then reset `startMessageID` to `markDeletePosition`.
+	if pc.hasSoughtByTime.CompareAndSwap(true, false) {
```

Comment:
This CAS operation seems confusing.

1. `hasSoughtByTime` is set with true in `internalSeekByTime`, so concurrent seek could happen. In Java client, it's prevented by `seekStatus.compareAndSet(SeekStatus.NOT_STARTED, SeekStatus.IN_PROGRESS)`.
2. Here it's changed from true to false, which means next time `hasNext` is called, it will skip the code block here.

Reply:
> Here it's changed from true to false, which means next time hasNext is called, it will skip the code block here.

Yes, After executing `seekByTime`, we only need to enter this method once, as it updates startMessageID to markDeletePosition. Now, it clearly knows what the messageId was from the last execution of seekByTime

Reply:
However, we need to add a new condition: when `&& lastDequeuedMsg == nil`. This is because the user might have received the message but hasn't called this method.